[cxx-interop] Conform C++ types with `std::hash` and `std::equal_to` specialization to `Swift.Hashable`

LukeXuan opened the pull request:
This pull request enables the clang importer to import `std::hash<T>` and `std::equal_to<T>` when available, and use them to conform to `Hashable` protocol.

## Changes
- the importer will now conform C++ types to `Hashable` if it finds a `std::hash` specialization and some source for `func ==`,
- `func ==` can come from either `operator ==` or `std::equal_to<T>`, with the former taking precedence,
- tests for the new functionality,

## Caveats
- In current implementation, `std::hash` and `std::equal_to` needs to be specialized context. In other words, one needs to type `using _ = std::hash<std::optional<std::string>>` in order to conform `std::optional<std::string>` to `Hashable`. Maybe we can try asking clang Sema to synthesize specializations for every C++ type, but that seems costly.
- Unlike C++ containers, `operator==` is prioritized before `std::equal_to`, since Swift doesn't distinguish these two.
- Derived point-wise `func ==` is not recognized by `getEqualEqualOperator`, and thus does not contribute to `Hashable` conformance for C++ types.

Resolves https://github.com/swiftlang/swift/issues/62045

LukeXuan committed:
[cxx-interop] use std::hash to implement Hashable

search for std::hash<T> when importing type T, and conform to Hashable
protocol automatically. The dependent Equatable protocol requires
existing operator==.

LukeXuan committed:
[cxx-interop] import std::equal_to when possible

Import std::equal_to<T> as the implementation for func == if,
- there is a specialization of std::equal_to<T>,
- there is no operator==,
and use it to conform to the Equatable protocol

LukeXuan committed:
[cxx-interop] introduce tests for Hashable

j-hui reviewed (commented):
Thank you for this contribution! I still need to review this more thoroughly but it's a welcome change. I left one comment about testing for now.

hnrklssn commented on lib/ClangImporter/ClangDerivedConformances.cpp:
```diff
@@ -1404,15 +1531,19 @@ void swift::conformToHashableIfNeeded(ClangImporter::Implementation &impl,
-  // we bail if we cannot find any operator==
-  if (!getEqualEqualOperator(decl))
-    return;
+  // if we cannot find std::equal_to<>, we can still try operator==
```
This comment makes it sound like `std::equal_to<>` is checked before `operator==`, but the comment further down makes it sound like the opposite.

LukeXuan replied:
Here's the problem:

In C++, `std::equal_to<T>`, used for `std::unordered_set<T>`, has a default implementation to call `operator ==` on type `T`. However, it is possible to provide an alternative implementation that doesn't agree with `operator ==`. For example,

```C++
#include<string>
#include<unordered_map>
#include<iostream>

using namespace std;

struct Key {
  int key;
  string comment;

  bool operator==(const Key &rhs) const {
    return key == rhs.key && comment == rhs.comment;
  }
};

template<>
struct std::hash<Key> {
    size_t operator() (const Key &k) const {
        return k.key;
    }
};

template<>
struct std::equal_to<Key> {
    bool operator() (const Key &lhs, const Key &rhs) const {
        return lhs.key == rhs.key;
    }
};

int main () {
  unordered_map<Key, string> m;
  m[Key{1, "hello"}] = "world";
  m[Key{1, "world"}] = "hello";

  cout << (Key{1, "hello"} == Key{1, "world"}) << endl; // expect to see 0
  
  cout << m[Key{1, "hello"}] << endl; // expect to see hello
  return 0;
}
```

As far as I know, Swift doesn't provide a way to use a different comparator for `==` and `Set`. So, we can achieve either,
- the same behavior for `==` in Swift and in C++ for C++ types, but potential different behavior for unordered containers,
- the same behavior for unordered containers, but different behavior in `==`.

My personal opinion is that the first choice is better.

hnrklssn replied:
This is an interesting aspect... Neither situation is great imo.
CC @egorzhdan what do you think?
Perhaps we want to emit some type of warning when `==` and `equal_to` diverge? One (heavy handed) way to handle it would be to error and force the user to select which of the implementations to import as `==`, using an annotation.

LukeXuan committed:
Update lib/ClangImporter/ClangDerivedConformances.cpp

Co-authored-by: Henrik G. Olsson <hnrklssn@gmail.com>

LukeXuan committed:
[cxx-interop] document std::equal_to trade-off

LukeXuan committed:
[cxx-interop] robust handling of import failure

add more checks when [impl.importDecl] returns null pointers, which
seems to happen for some std types on MacOS tests

LukeXuan committed:
[cxx-interop] use -typecheck -verify for tests

LukeXuan commented on test/Interop/Cxx/stdlib/use-std-hash-typechecker.swift:
```diff
@@ -0,0 +1,13 @@
+// RUN: not %target-swift-frontend %s -typecheck -I %S/Inputs -cxx-interoperability-mode=default -diagnostic-style llvm 2>&1 | %FileCheck %s
```
I have pushed a commit that changed this. I'm using `%target-typecheck-verify-swift` which seems to be macro for `-typecheck -verify`.

hnrklssn commented on test/Interop/Cxx/stdlib/use-std-hash-typechecker.swift:
```diff
@@ -1,13 +1,10 @@
-// RUN: not %target-swift-frontend %s -typecheck -I %S/Inputs -cxx-interoperability-mode=default -diagnostic-style llvm 2>&1 | %FileCheck %s
+// RUN: %target-typecheck-verify-swift %s -I %S/Inputs -cxx-interoperability-mode=default -diagnostic-style llvm
```
I don't think `-diagnostic-style llvm` makes a difference when using `-verify`

LukeXuan committed:
[cxx-interop] use ASSERT instead of assert

Co-authored-by: Henrik G. Olsson <hnrklssn@gmail.com>

LukeXuan committed:
[cxx-interop] remove useless argument in test file

hnrklssn commented:
> CI seems to consistently fail on Linux platform, although my local environment (Fedora 43) works fine. Is [this](https://github.com/swiftlang/swift-docker/blob/main/swift-ci/main/ubuntu/22.04/Dockerfile) the image used for CI? If so I can try debug locally.

Yup, that's the one. Note that the test passed in CI for check-swift-all-linux-x86_64, but failed for check-swift-all-optimize-linux-x86_64.

edit: I suggest editing the dockerfile to install `vim` or your editor of choice, because it doesn't contain any editor by default.

LukeXuan committed:
[cxx-interop] replace std::string with std::bitset

std::string receives its own custom extension on Swift side, using it to
test std::hash importation is pointless. std::bitset makes more sense
here.

LukeXuan commented:
Hmmm, I don't have access to MacOS or Windows machine. Judging from the error log, it's caused by not including `<bitset>`. If you have access to MacOS environment, can you verify if that's the case?

hnrklssn commented on test/Interop/Cxx/stdlib/use-std-hash-typechecker.swift:
```diff
@@ -0,0 +1,10 @@
+// RUN: %target-typecheck-verify-swift %s -I %S/Inputs -cxx-interoperability-mode=default
```
How does this pass on Linux? On macOS it errors with:
```
[2025-12-02T05:32:47.345Z] <unknown>:0: error: duplicate input file '/Users/ec2-user/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/swift/test/Interop/Cxx/stdlib/use-std-hash-typechecker.swift'
```
since `%target-typecheck-verify-swift` includes `%s`

LukeXuan committed:
[cxx-interop] fix test cases

adds missing header includes, and removes duplicate %s

j-hui commented on lib/ClangImporter/ClangDerivedConformances.cpp:
```diff
@@ -1364,3 +1364,218 @@ void swift::conformToCxxSpanIfNeeded(ClangImporter::Implementation &impl,
+lookupAndSpecializeFunctionObject(ClangImporter::Implementation &impl,
+                                  const clang::CXXRecordDecl *clangDecl,
+                                  std::string name) {
+  ASSERT(clangDecl);
```
nit: I don't think this is necessary: this is a `static` helper function whose caller already asserts this.

```suggestion
```

j-hui commented on lib/ClangImporter/ClangDerivedConformances.cpp:
```diff
@@ -1364,3 +1364,218 @@ void swift::conformToCxxSpanIfNeeded(ClangImporter::Implementation &impl,
+      stdNS->lookup(declName).find_first<clang::ClassTemplateDecl>();
+
+  if (!tmplDecl)
+    return nullptr;
```
Does this lookup need to be repeated (twice) upon every import? This seems like the kind of thing that should be memoized by `ClangImporter`, somewhere, since it is not dependent on the `clangDecl`, which is only used later to find the specialization.

LukeXuan replied:
I'm confused by the "(twice)" part, but I assumed you mean the `ClassTemplateDecl` for `name`? I think yes it should be memorized by the importer.

Though I'm not familiar with that part of the code structure. Which class would be the best place to store these lookup information? Is it the `ClangImporter::Implementation` class?

j-hui replied:
> I'm confused by the "(twice)" part, but I assumed you mean the ClassTemplateDecl for name? I think yes it should be memorized by the importer.

Yeah, that's what I meant. I'm not sure if there's any cacheing on the Clang side that makes `stdNS->lookup()` efficient but I don't think it should happen `2 * N` times if we import `N` structs (once for `std::hash` and once for `std::equal_to`).

And yeah, the "easiest" thing to do is to stuff it in `ClangImporter::Implementation` as a member variable, and you can lazily initialize it or something like that. Though that struct is a bit cluttered, so adding another field might be bad from an organization standpoint. @egorzhdan what are your thoughts here?

j-hui commented on lib/ClangImporter/ClangDerivedConformances.cpp:
```diff
@@ -1364,3 +1364,218 @@ void swift::conformToCxxSpanIfNeeded(ClangImporter::Implementation &impl,
+    declContext = declContext->getParent();
+  }
+
+  clang::CXXMethodDecl *methodDecl = nullptr;
```
Sorry if I'm missed this from `conformToHashableIfNeeded`, but is it ever possible for `methodDecl` to never be assigned? What guarantees the presence of a suitable `operator ()`?

j-hui replied:
Perhaps this assumption is only valid if this helper function is called with `std::equal_to`, but I think that could/should be made clearer.

LukeXuan replied:
The intention is that this function is invoked with a function object `funcDecl` (one with `operator()`). I should put an `ASSERT(methodDecl)` after the search.

j-hui commented on lib/ClangImporter/ClangDerivedConformances.cpp:
```diff
@@ -1364,3 +1364,218 @@ void swift::conformToCxxSpanIfNeeded(ClangImporter::Implementation &impl,
+  for (auto &&candidate : funcDecl->methods()) {
+    if (candidate->getOverloadedOperator() ==
+            clang::OverloadedOperatorKind::OO_Call &&
+        candidate->param_size() == 2) {
```
What guarantees that there is exactly one suitable `operator ()`?

j-hui commented on lib/ClangImporter/ClangDerivedConformances.cpp:
```diff
@@ -1364,3 +1364,218 @@ void swift::conformToCxxSpanIfNeeded(ClangImporter::Implementation &impl,
+      clangSema.BuildCallExpr(nullptr, synthCtorExpr, clang::SourceLocation(),
+                              args, clang::SourceLocation());
+
+  if (!call.isUsable())
```
What are valid reasons the call expression is unusable? If there are none, then maybe `ASSERT()` like you did above?

j-hui reviewed (commented):
Thanks for this patch! I left some feedback, mostly clarification questions, but also some suggestions for the implementation.

LukeXuan commented on test/Interop/Cxx/stdlib/Inputs/std-hash.h:
```diff
@@ -0,0 +1,64 @@
+  int value;
+  std::string comment;
+
+  bool operator==(const A &rhs) const { return value == rhs.value; }
```
I would argue it should be tested here https://github.com/swiftlang/swift/tree/main/test/Interop/Cxx/operators/Inputs instead of this file, and I think they are tested already? Of course, the assumption is that `getEqualEqualOperator` will return the function in all cases.

j-hui commented on lib/ClangImporter/ClangDerivedConformances.cpp:
```diff
@@ -1364,3 +1364,217 @@ void swift::conformToCxxSpanIfNeeded(ClangImporter::Implementation &impl,
+
+  if (auto *stdHashStruct = dyn_cast<StructDecl>(stdHash)) {
+    FuncDecl *hashFunc = synthesizer.makeHashFunc(decl, stdHashStruct);
+    decl->addMember(hashFunc);
```
I'm wondering whether we should wait to import `std::hash`, synthesize the hash function, and add this member to `decl` if and only if we synthesize the conformance? Or is it intentional that we add a hash method even if we can't find a suitable `==(_:_:)`?

j-hui reviewed (commented):
This patch is looking great so far! I left some more comments, mostly concerns about whether this will trigger new compiler crashes, errors, or miscompilations. Sorry to increase the test burden, but I really appreciate the effort so far!

LukeXuan committed:
address most comments

LukeXuan commented:
Sorry I ran out of steam in the past few month, but I'm back!

I tried to address most of the previous comments, except for [this one](https://github.com/swiftlang/swift/pull/85283#discussion_r2594513144) which I need to look into retroactive annotation closer to understand what to do. In the meanwhile, here's the high level summary of changes so you can take a look:

1. I attempted to hoist most of the lookup earlier and defer synthesis later. Especially around finding the `std::equal_to` and `std::hash` specializations,
2. For specialization discovery, I also check for the inhabitance of the intended method by type and qualifiers, and stops the process if either one is not found.
3. I attempted to decouple the logic for Equatable and Hashable conformance,
4. I added a warning to the compiler to highlight the fact that C++ equality test and Swift equality test for hash-based data structures might be different. I'm not sure if that's the place/the utility I should use, please advise!

The code infrastructure surrounding my pull request also changed significantly, so it's probably worthwhile to review the PR as a whole again.

LukeXuan committed:
decouple specialization lookup and method lookup

also makes sure that hasher synthesis always succeeds by checking default
constructor in advance

LukeXuan commented:
uhhh, it seems like I made some mistake when pulling from upstream, I will try to clean it up...

LukeXuan commented:
OK, I think I cleaned up the commit history. If any of the cxx-interop team can help me remove all the bogus review request I would appreciate!

j-hui commented on include/swift/AST/DiagnosticsClangImporter.def:
```diff
@@ -455,6 +455,11 @@ ERROR(conflicting_swift_attrs, none,
 
+WARNING(mismatched_equatable_semantics, none,
+        "imported C++ module exposes both equal operator and std::equal_to<>"
+        "for %0, prioritizing the equal operator. ",
```
```suggestion
        "for %0, prioritizing the equal operator.",
```

j-hui commented:
Thanks for revisiting this PR @LukeXuan ! I was just thinking about our lack of hashable conformances the other day. I removed all the stray reviewers and will do a thorough re-review myself.

CC @patrykstefanski , who has also been thinking about revamping some of the operator lookup logic to lean more on Clang-side lookup logic, rather than relying on Swift/ClangImporter's currently lazy import behavior.

j-hui commented on lib/ClangImporter/ClangDerivedConformances.cpp:
```diff
@@ -1592,9 +1596,318 @@ static void conformToCxxSpan(ClangImporter::Implementation &impl,
+  auto &clangCtx = impl.getClangASTContext();
+  auto &clangSema = impl.getClangSema();
+
+  /// 2. constructor the operator
```
This seems like a stray comment that lacks a corresponding `/// 1.`

I would prefer if this and the subsequent numbered comments are removed, since the clang Sema calls roughly describe what's going on (albeit with a quite a bit of line noise).

j-hui commented on lib/ClangImporter/ClangDerivedConformances.cpp:
```diff
@@ -1592,9 +1596,318 @@ static void conformToCxxSpan(ClangImporter::Implementation &impl,
+static bool synthesizeOperatorWithCxxFunctionObject(
+    ClangImporter::Implementation &impl, const clang::CXXRecordDecl *classDecl,
+    const clang::CXXRecordDecl *fobjDecl,
+    clang::CXXConstructorDecl *ctorDecl, // why not const??
```
This comment should probably be removed, but I think the reason is `clang::Sema::BuildCXXConstructExpr()` needs to internally mark the function as used (i.e., `clang::Sema::MarkFunctionReferenced()`), which mutates the decl.

Another reason might be that `BuildCXXConstructExpr()` may need to lazily synthesize the definition of an implicit default constructor, but I'm less sure about that.

j-hui commented on lib/ClangImporter/ClangDerivedConformances.cpp:
```diff
@@ -1592,9 +1596,318 @@ static void conformToCxxSpan(ClangImporter::Implementation &impl,
+  const char *opSpelling = clang::getOperatorSpelling(opKind);
+  auto declName = clang::DeclarationName(&clangCtx.Idents.get(opSpelling));
+
+  // Determine the Clang decl context where the new operator function will be
+  // created. We use the translation unit as the decl context of the new
+  // operator, otherwise, the operator might get imported as a static member
+  // function of a different type (e.g. an operator declared inside of a C++
+  // namespace would get imported as a member function of a Swift enum), which
+  // would make the operator un-discoverable to Swift name lookup.
+  auto declContext =
+      const_cast<clang::CXXRecordDecl *>(classDecl)->getDeclContext();
+  while (!declContext->isTranslationUnit()) {
+    declContext = declContext->getParent();
+  }
```
This comment and logic seems duplicated from `synthesizeCXXOperator`. I'm wondering if there's any opportunity to merge these?

j-hui commented on lib/ClangImporter/ClangDerivedConformances.cpp:
```diff
@@ -1592,9 +1596,318 @@ static void conformToCxxSpan(ClangImporter::Implementation &impl,
+  clang::QualType returnTy = methodDecl->getReturnType();
+  clang::QualType lhsTy = clangCtx.getTypeDeclType(classDecl),
+                  rhsTy = clangCtx.getTypeDeclType(classDecl);
+  clang::FunctionProtoType::ExtProtoInfo epi;
+  if (const auto *protoTy =
+          methodDecl->getType()->getAs<clang::FunctionProtoType>()) {
+    epi = protoTy->getExtProtoInfo();
+  } else {
+    ASSERT(false && "unexpected error when copying ExtProtoInfo");
+  }
```
IMO it is cleaner to do:

```suggestion
  const auto *methodProtoTy =
          methodDecl->getType()->getAs<clang::FunctionProtoType>();
  ASSERT(methodProtoTy && "method should always have a function prototype");
```

And then inline `methodProtoTy->getExtProtoInfo()` later.

j-hui commented on lib/ClangImporter/ClangDerivedConformances.cpp:
```diff
@@ -1592,9 +1596,318 @@ static void conformToCxxSpan(ClangImporter::Implementation &impl,
 void swift::deriveAutomaticCxxConformances(
-    ClangImporter::Implementation &Impl, NominalTypeDecl *result,
-    const clang::CXXRecordDecl *clangDecl) {
+    ClangImporter::Implementation &Impl, SwiftDeclSynthesizer &synthesizer,
```
This is just a code organization nit, but IMO we shouldn't need to pull in `SwiftDeclSynthesizer` here (which is just a wrapper over `ClangImporter::Implementation &ImporterImpl`, which `makeHashFunc` only uses to obtain the `swift::ASTContext &`).

I would prefer to pull `makeHashFunc` into this file as a `static` helper function, to simplify the code organization, since it's not used anywhere else.

j-hui commented on lib/ClangImporter/ImporterImpl.h:
```diff
@@ -2027,6 +2027,12 @@ class LLVM_LIBRARY_VISIBILITY ClangImporter::Implementation
+  std::optional<const clang::NamespaceDecl *> stdNS;
+
+public:
+  const clang::NamespaceDecl *getCxxStdNamespace();
```
I don't think this cache is necessary. `clang::Sema::getStdNamespace()` already uses a `LazyDeclPtr` internally, which iiuc does the caching for us already.

j-hui commented on lib/ClangImporter/ClangDerivedConformances.cpp:
```diff
@@ -1592,9 +1596,318 @@ static void conformToCxxSpan(ClangImporter::Implementation &impl,
+      FuncDecl *hashFunc = synthesizer.makeHashFunc(decl, stdHashStruct);
+      decl->addMember(hashFunc);
+    } else {
+      ASSERT(false);
```
Instead of `ASSERT(false)`, prefer `ABORT("message")` or `ABORT([&](auto &out) { ... })` for better crash diagnostics. `ASSERT(stdHashStruct && "message")` works too.

j-hui commented on lib/ClangImporter/ClangDerivedConformances.cpp:
```diff
@@ -1592,9 +1596,318 @@ static void conformToCxxSpan(ClangImporter::Implementation &impl,
+// Attempt at decoupling the std::equal_to<> conformance from std::hash<>
+// conformance. Because we don't want unnecessary importation, the actual import
+// is delayed as a thunk.
+static std::optional<std::function<void ()>> conformToEquatableIfNeeded(
```
I don't think this thunking is actually necessary. Automatically deriving a conformance to equatable actually seems quite useful and probably something we'd want even for things that don't conform to `std::hash<>`.

And even if we only want the conformance as part of `std::hash<>`, I still think it's cleaner if this didn't return a function object.

j-hui reviewed (commented):
I left some various bits of feedback.

I wanted to reflect on this aspect of the high-level design as part of the main review comment:

```
  // we have two possible sources for [==(_:_:)]
  // 1. [std::equal_to<T>]
  // 2. [operator==(T, T)]
  // The current implementation prioritize 2 over 1, which as the following
  // effects,
  // - the behavior of [==(_:_:)] operator in Swift will match that of
  // [operator==(T, T)] in C++, if both exist,
  // - the behavior of unordered containers (such as [Set]) in Swift will not
  // match their equivalences ([unordered_set]) in C++, if the C++ type has a
  // custom [std::equal_to<T>] specialization that does not agree with
  // [operator==(T, T)]
```

I actually think we should favor `std::equal_to` over `operator==`, because, at least to me, an explicit template instance of `std::equal_to` is a stronger signal for the _semantic_ equality of two objects, vs `operator==` which might only be defined a certain way as a matter of syntactic convenience.

That said, I think favoring `std::equal_to` may have implications on how we are currently importing `operator==`, since we want to avoid the situation where the type checker gets upset by two conflicting implementations of `func ==`, one imported from `operator==` and one imported from `std::equal_to`. But in the common case, the default implementation of `std::equal_to::operator()` will just use `operator==`, so the behavioral impact should hopefully be minimal.

As a result, the overall logic of this PR that I'm proposing will look something like:

```python
if has explicit std::equal_to<T>:
  conform T to Equatable using std::equal_to<T>::operator()
else if exists operator==(T, T):
  conform T to Equatable using operator==(T, T)
  # equivalently, conform using implicit instantiation of std::equal_to<T>

if T was conformed to Equatable:
  if has explicit std::hash<T>:
    # more specifically std::hash<T> is "enabled";
    # see also: https://en.cppreference.com/cpp/utility/hash
    conform T to Hashable using std::hash<T>::operator()
```

Another consideration here is whether to instantiate the `std::equal_to<T>` instance when one does not exist. I chose not to do this for `std::iterator_traits` because the alternative there is to just do a qualified look up for the `iterator_tag` member. `getEqualEqualOperator()` is quite a bit more expensive since it attempts several lookups, plus importing `operator==` wraps `__operatorEqualsEquals` with a synthesized `func ==`, so the relative cost of just instantiating `std::equal_to<T>` doesn't seem as bad. CC @patrykstefanski and @Xazax-hun who had both worked on this recently.

(Side note: it would be really great if all this eager look up and synthesis could be lazily triggered by the conformance checker, but I don't think that's possible yet as of the last time I looked into it.)

Anyway, @LukeXuan I hope that all makes sense. Sorry I didn't give this opinionated feedback in December when you first opened this PR: I'd since worked on the conformance checking for `std::iterator_traits` and that shaped a lot of my views about how I think we should be deriving conformances from the C++ stdlib's trait templates. I'm cognizant that my comments may lead to a lot of work for your first contribution to Swift, so please don't hesitate to ping me for help.

LukeXuan commented:
> I actually think we should favor std::equal_to over operator==, because, at least to me, an explicit template instance of std::equal_to is a stronger signal for the semantic equality of two objects, vs operator== which might only be defined a certain way as a matter of syntactic convenience.

I recall that I avoided doing this way for two reasons:
1. I need to adventure into existing importation logic to modify and de-prioritize those, which I wasn't very comfortable with,
2. It would cause backward compatibility issues? If existing Swift code already depends on the automatically imported `operator==` implementation[1], prioritizing `std::equal_to<>` would break it.

> `getEqualEqualOperator()` is quite a bit more expensive since it attempts several lookups

I have to point out that [here](https://github.com/swiftlang/swift/pull/85283/changes#diff-29cfd49b447512a0f61a15554bb0052146d9293af340aed5bf854fbb232d108aR1722) I'm also using a somewhat expensive lookup to locate the `operator()` for both `std::equal_to` and `std::hash`, as the simple member lookup will fail for some STL types. I'm not sure if it's still cheaper than `getEqualEqualOperator`.

Also regarding cost, I'm also forcing instantiation [here](https://github.com/swiftlang/swift/pull/85283/changes#diff-29cfd49b447512a0f61a15554bb0052146d9293af340aed5bf854fbb232d108aR1761), because otherwise the `hash`/`equal_to` instances cannot be found for STL types. Though I'm not sure what you meant in `explicit hash/equal_to`, maybe it's acceptable to not import these STL hash implementation unless the user had some annotations in the C++ header?

[1] For what I understand, Swift do not need `Equatable` conformance to be able to write `x == y`, all it needs is `func==` somewhere in the context.
